File: 2024/03/main.py
import re
from sys import stdin

total = 0
p2_total = 0
enabled = True

for match in re.finditer(r"(?:mul\((\d+),(\d+)\))|don\'t\(\)|do\(\)", stdin.read()):
    match match.group(0):
        case "do()":
            enabled = True

        case "don't()":
            enabled = False

        case _:
            mult = int(match.group(1)) * int(match.group(2))
            total += mult

            if enabled:
                p2_total += mult

print(total, p2_total)

File: 2024/03/main.js
import * as readline from 'readline';

const rl = readline.createInterface({
  input: process.stdin,
})

let p1total = 0;
let p2total = 0;
let enabled = true;

rl.on('line', (line) => {
  line.matchAll(/(?:mul\((\d{1,3}),(\d{1,3})\))|don\'t\(\)|do\(\)/gm).forEach(match => {
    switch (match[0]) {
      case "do()":
        enabled = true;
        break;

      case "don't()":
        enabled = false;
        break;

      default:
        p1total += match[1] * match[2];

        if (enabled) {
          p2total += match[1] * match[2];
        }
    }
  });
})

rl.on('close', () => {
  console.log(p1total, p2total);
})
